metaflow/plugins/tagging_cli.py

```diff
@@ -0,0 +1,424 @@
+            msg += "You can choose any of the following tags associated with *%s* " % run_name + \
+                   "to use with option *--namespace*\n"
+            msg += "\n".join(["\t*%s*" % tag for tag in user_tags])
+
+        raise CommandException(msg)
+
+    return metaflow_client
+
+
+def list_runs(obj, user_namespace):
+    namespace(user_namespace)
+    metaflow_client = Flow(pathspec=obj.flow.name)
+    return metaflow_client.runs()
+
+
+@click.group()
+def cli():
+    pass
+
+
+@cli.group(help='Commands related to Metaflow tagging.')
+def tag():
+    pass
+
+
+@tag.command('add',
+             help='Add tags to a run.')
+@click.option('--run-id',
+              required=False,  # set False here so we can throw a better error message
+              default=None,
+              type=str,
+              help="Run ID of the specific run to tag. [required]")
+@click.option('--namespace',
+              'user_namespace',
+              required=False,
+              default=None,
+              type=str,
+              help="Change namespace from the default (your username) to "
+                   "the specified tag.")
+@click.argument('tags',
+                required=True,
+                type=str,
+                nargs=-1)
+@click.pass_obj
+def add(obj, run_id, user_namespace, tags):
+    user_namespace = resolve_identity() if user_namespace is None else user_namespace
+    metaflow_run_client = obtain_metaflow_run_client(obj,
+                                                     run_id, user_namespace)
+
+    metaflow_run_client.add_tags(tags)
+
+    obj.echo("Operation successful. New tags:", err=False)
+    print_tags(obj, metaflow_run_client)
+
+
+@tag.command('remove',
+             help='Remove tags from a run.')
+@click.option('--run-id',
+              required=False,  # set False here so we can throw a better error message
+              default=None,
+              type=str,
+              help="Run ID of the specific run to tag.  [required]")
+@click.option('--namespace',
+              'user_namespace',
+              required=False,
+              default=None,
+              type=str,
+              help="Change namespace from the default (your username) to "
+                   "the specified tag.")
+@click.argument('tags',
+                required=True,
+                type=str,
+                nargs=-1)
+@click.pass_obj
+def remove(obj, run_id, user_namespace, tags):
+    user_namespace = resolve_identity() if user_namespace is None else user_namespace
+    metaflow_run_client = obtain_metaflow_run_client(obj,
+                                                     run_id, user_namespace)
+
+    metaflow_run_client.remove_tags(tags)
+
+    obj.echo("Operation successful. New tags:")
+    print_tags(obj, metaflow_run_client)
+
+
+@tag.command('replace',
+             help='Replace tags of a run.')
+@click.option('--run-id',
+              required=False,  # set False here so we can throw a better error message
```

Comment:
Why can't we set this to required? It will simplify the logic in the first 200 lines of this file tremendously.

Reply:
Are you talking about all the exceptions handling at the top? I think the idea was to try to be friendly to the user. @jasonge27 might have more insights (I copied a lot of the code).

Reply:
Right, simply want to be more user-friendly. If we set it to False, and if the user does not specify run-id, `click` will throw an error with the help message